reframe/frontend/executors/policies.py

```diff
@@ -224,19 +275,50 @@ def runcase(self, case):
             self._failall(e)
             raise
 
+    def _cleanup_all(self):
+        for task in self._retired_tasks:
+            if task.ref_count == 0:
+                task.cleanup(not self.keep_stage_files)
+
+        # Remove cleaned up tests
+        self._retired_tasks[:] = [
+            t for t in self._retired_tasks if t.ref_count]
+
     def _poll_tasks(self):
         '''Update the counts of running checks per partition.'''
         getlogger().debug('updating counts for running test cases')
         getlogger().debug('polling %s task(s)' % len(self._running_tasks))
         for t in self._running_tasks:
             t.poll()
 
+    def _setup_all(self):
+        tasks_to_remove = []
```

Comment:
This name is a bit confusing and the statement at the end that you remove them from the waiting list. I would write it as follows:

```python
still_waiting = []
for stuff:
    if some_stuff:
        ...
    elif other_stuff:
        ...
    else:
        still_waiting.append(task)

self._waiting_tasks[:] = still_waiting
```